File: teambuilder/allocation/constraints.py
# This file defines classes for representing constraints
# (besides group size) in an allocation.

import math


# Calculate the mean value of some property in a team
def team_mean(team, student_info, field):
	return sum(student_info[student][field] for student in team) / len(team)

# Calculate the standard deviation of some property in a team
def team_stdev(team, student_info, field):
	team_average = team_mean(team, student_info, field)
	sqdiffs = sum((student_info[student][field] - team_average)**2 for student in team)
	return math.sqrt(sqdiffs / (len(team)-1))

# Return all members of a team who do / do not meet some condition
def team_filter(team, student_info, field, with_bool, function):
	return [student for student in team if function(student_info[student][field]) ^ (not with_bool)]


# Simple class for working with ranges in constraints
class BXY:
	def __init__(self, lower, upper):
		self.lower = lower
		self.upper = upper
	
	# returns the width of the interval represented by the BXY
	def width(self):
		return self.upper - self.lower + 1
	
	# returns how far outside the BXY a value is, or 0 if the value is inside
	def distance(self, value):
		return max(0, self.lower - value, value - self.upper)
	
	# returns how many widths outside of the BXY a value is
	def scaled_distance(self, value):
		return self.distance(value) / self.width()
	
	# returns how far inside the BXY a value is, or 0 if the value is outside
	def inclusion(self, value):
		return max(0, min(value - self.lower + 1, self.upper - value + 1))
	
	# returns what fraction of a radii inside of the BXY a value is
	def scaled_inclusion(self, value):
		radius = math.ceil(self.width() / 2)
		return self.inclusion(value) / radius
	
	def __contains__(self, item):
		return self.lower <= item <= self.upper
	
	def __len__(self):
		return 2
	
	def __getitem__(self, key):
		if key == 0:
			return self.lower
		elif key == 1:
			return self.upper
		else:
			raise IndexError


# ABC for all constraints
class Constraint:
	def __init__(self, name, field, constraint_type, priority):
		self.name = name
		self.field = field
		self.constraint_type = constraint_type
		self.priority = priority
	
	# Returns a "badness" score for the tested team
	def evaluate(self, team, student_info):
		raise NotImplementedError


# A constraint that controls the number of members per-team which match an integer-based condition.
class IntegerCountConstraint(Constraint):
	should_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	shouldnt_tune = 1.0
	
	# Each team <should/shouldn’t> have <BXY> members <with/without> [age] <BXY>.
	#                   |                 \-----------------\  |             |
	#                   \------------------------- \        |  \--------\    \------\
	#                                              V        V           V           V
	def __init__(self, name, field, priority, should_bool, count_bxy, with_bool, value_bxy):
		super().__init__(name, field, "integer", priority)
		self.should_bool = should_bool
		self.count_bxy = count_bxy
		self.with_bool = with_bool
		self.value_bxy = value_bxy
	
	def evaluate(self, team, student_info):
		count = len(team_filter(team, student_info, self.field, self.with_bool, lambda x: x in self.value_bxy))
		if self.should_bool:
			return self.should_tune * self.priority * self.count_bxy.scaled_distance(count)
		else:
			return self.shouldnt_tune * self.priority * self.count_bxy.scaled_inclusion(count)


# A constraint that controls the team-wide average of some property.
class IntegerAverageConstraint(Constraint):
	should_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	shouldnt_tune = 1.0
	
	# Each team <should/shouldn’t> have an average [age] <BXY>.
	#                   |                                  |
	#                   \--------------------------\       \----\
	#                                              V            V
	def __init__(self, name, field, priority, should_bool, average_bxy):
		super().__init__(name, field, "integer", priority)
		self.should_bool = should_bool
		self.average_bxy = average_bxy
	
	def evaluate(self, team, student_info):
		mean = team_mean(team, student_info, self.field)
		if self.should_bool:
			return self.should_tune * self.priority * self.average_bxy.scaled_distance(mean)
		else:
			return self.shouldnt_tune * self.priority * self.average_bxy.scaled_inclusion(mean)


# A constraint that controls the team-wide average of some property.
class IntegerSimilarityConstraint(Constraint):
	tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	
	# Each team should have <similar/diverse> [age].
	#                               |
	#                               \--------------\
	#                                              V
	def __init__(self, name, field, priority, similar_bool):
		super().__init__(name, field, "integer", priority)
		self.similar_bool = similar_bool
	
	def evaluate(self, team, student_info):
		stddev = team_stdev(team, student_info, self.field)
		if self.similar_bool:
			return self.tune * self.priority * stddev
		else:
			min_val = min(student_info[student][self.field] for student in student_info)
			max_val = max(student_info[student][self.field] for student in student_info)
			max_std = (max_val - min_val) / math.sqrt(2)
			return self.tune * self.priority * (max_std - stddev)


# A constraint that controls how close teams are to the global average of some property.
class IntegerGlobalAverageConstraint(Constraint):
	tune = 1.0	# tuning value to match influence of different constraints (with the same priority).
	
	# All teams should have similar average [age].
	def __init__(self, name, field, priority):
		super().__init__(name, field, "integer", priority)
	
	def evaluate(self, team, student_info):
		global_mean = team_mean(student_info.keys(), student_info, self.field)
		mean = team_mean(team, student_info, self.field)
		return self.tune * self.priority * abs(global_mean - mean)


# A constraint that controls the number of members per-team which selected a particular option
# "candidates" is a list of valid answers to the question
class OptionCountConstraint(Constraint):
	should_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	shouldnt_tune = 1.0
	
	# Each team <should/shouldn’t> have <BXY> members <with/without> [degree] <VALUE>.
	#                   |                 \-----------------\   |                |
	#                   \------------------------- \        |   \--------\       \--\
	#                                              V        V            V          V
	def __init__(self, name, field, priority, should_bool, count_bxy, with_bool, selection, candidates):
		super().__init__(name, field, "option", priority)
		self.should_bool = should_bool
		self.count_bxy = count_bxy
		self.with_bool = with_bool
		self.selection = selection
		self.candidates = candidates
	
	def evaluate(self, team, student_info):
		count = len(team_filter(team, student_info, self.field, self.with_bool, lambda x: x == self.selection))
		if self.should_bool:
			return self.should_tune * self.priority * self.count_bxy.scaled_distance(count)
		else:
			return self.shouldnt_tune * self.priority * self.count_bxy.scaled_inclusion(count)


# A constraint which requires members of a team to be more/less similar
# to each other in their answers to an "option" (select one of) question
# "candidates" is a list of valid answers to the question
class OptionSimilarityConstraint(Constraint):
	similar_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	diverse_tune = 1.0
	
	# Each team should have <similar/diverse> [degree].
	#                              |
	#                              \---------------\
	#                                              V
	def __init__(self, name, field, priority, similar_bool, candidates):
		super().__init__(name, field, "option", priority)
		self.similar_bool = similar_bool
		self.candidates = candidates
	
	def evaluate(self, team, student_info):
		votes = {candidate:0 for candidate in self.candidates}
		
		# Collect votes
		for student in team:
			choice = student_info[student][self.field]
			votes[choice] += 1
		
		if self.similar_bool:
			return self.similar_tune * self.priority * (len(team) - max(votes.values()))
		else:
			return self.diverse_tune * self.priority * (len(team)/len(self.candidates) - min(votes.values()))


# A constraint that controls the number of members per-team which selected a particular choice
# "candidates" is a list of valid answers to the question
class SubsetCountConstraint(Constraint):
	should_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	shouldnt_tune = 1.0
	
	# Each team <should/shouldn’t> have <BXY> members <with/without> [interests] <VALUE>.
	#                   |                 \-----------------\   |                   |
	#                   \------------------------- \        |   \--------\          |
	#                                              V        V            V          V
	def __init__(self, name, field, priority, should_bool, count_bxy, with_bool, selection, candidates):
		super().__init__(name, field, "subset", priority)
		self.should_bool = should_bool
		self.count_bxy = count_bxy
		self.with_bool = with_bool
		self.selection = selection
		self.candidates = candidates
	
	def evaluate(self, team, student_info):
		count = len(team_filter(team, student_info, self.field, self.with_bool, lambda x: self.selection in x))
		if self.should_bool:
			return self.should_tune * self.priority * self.count_bxy.scaled_distance(count)
		else:
			return self.shouldnt_tune * self.priority * self.count_bxy.scaled_inclusion(count)


# A constraint requires members of a team to be more/less similar to each other
# in their answers to a "subset" (select multiple values) question
class SubsetSimilarityConstraint(Constraint):
	similar_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	diverse_tune = 1.0
	
	# Each team should have <similar/diverse> [interests].
	#                             |
	#                             \----------------\
	#                                              V
	def __init__(self, name, field, priority, similar_bool, candidates):
		super().__init__(name, field, "subset", priority)
		self.similar_bool = similar_bool
		self.candidates = candidates
	
	def evaluate(self, team, student_info):
		votes = {candidate:0 for candidate in self.candidates}
		voter_count = 0
		
		# Collect votes
		for student in team:
			num_votes = len(student_info[student][self.field])
			if num_votes > 0:
				voter_count += 1
				for candidate in student_info[student][self.field]:
					votes[candidate] += 1/num_votes
		
		if self.similar_bool:
			return self.similar_tune * self.priority * (voter_count - max(votes.values()))
		else:
			return self.diverse_tune * self.priority * (voter_count/len(self.candidates) - min(votes.values()))


# A constraint that controls the number of members per-team with some boolean criteria
# "candidates" is a list of valid answers to the question
class BooleanCountConstraint(Constraint):
	should_tune = 1.0	# tuning values to match influence of different constraints (with the same priority).
	shouldnt_tune = 1.0
	
	# Each team <should/shouldn’t> have <BXY> members <with/without> [postgraduate].
	#                   |                 \-----------------\   |
	#                   \------------------------- \        |   \--------\
	#                                              V        V            V
	def __init__(self, name, field, priority, should_bool, count_bxy, with_bool):
		super().__init__(name, field, "boolean", priority)
		self.should_bool = should_bool
		self.count_bxy = count_bxy
		self.with_bool = with_bool
	
	def evaluate(self, team, student_info):
		count = len(team_filter(team, student_info, self.field, self.with_bool, lambda x: x))
		if self.should_bool:
			return self.should_tune * self.priority * self.count_bxy.scaled_distance(count)
		else:
			return self.shouldnt_tune * self.priority * self.count_bxy.scaled_inclusion(count)

File: teambuilder/teambuilder/static/js/constraints.js
// This file contains definitions of constraints that are
// able to be used by the allocator. Note that the error
// checking of these constraints is limited to semantic
// correctness only. This is to allow for modularity.

// VALIDATION METHODS -----------------------------------------------
function isBoolean(value) {
    return typeof value === "boolean";
}

function isNumber(value) {
    return typeof value === "number" && isFinite(value);
}

function isString(value) {
    return typeof value === "string" || value instanceof String;
}

function isArray(value) {
    return value && typeof value === "object" && value.constructor === Array;
}

// CONSTRAINTS ------------------------------------------------------
var createObject = function(o) {
    function F() {}
    F.prototype = o
    return new F();
};

function IntegerCountConstraint(shouldBool, countMin, countMax,
    withBool, field, fieldMin, fieldMax) {
    // Each team <should/shouldn't> have between <min>
    // and <max> members <with/without> <field> between
    // <min> and <max>.

    var isError = false;
    var errMsg = "";

    if (!isNumber(countMin) || !isNumber(countMax) || countMin > countMax) {
        errMsg = "Invalid count constraint bounds";
        isError = true;
    } else if (!isBoolean(shouldBool) || !isBoolean(withBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    } else if (!isNumber(fieldMin) || !isNumber(fieldMax) || fieldMin > fieldMax) {
        errMsg = "Invalid count constraint values";
        isError = true;
    }

    var params = {
        "constr_type" : "IntegerCountConstraint",
        "name" : field + " constraint",
        "should_bool" : (shouldBool == 'true'),
        "count_bxy" : [countMin, countMax],
        "with_bool" : (withBool == 'true'),
        "field" : field,
        "value_bxy" : [fieldMin, fieldMax],
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        console.log(JSON.stringify(params));
        console.log(JSON.parse(JSON.stringify(params)));
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["should_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "between ";
        sentence += "<strong>" + params["count_bxy"][0] + "</strong> ";
        sentence += "and "
        sentence += "<strong>" + params["count_bxy"][1] + "</strong> ";
        sentence += "members ";
        if (params["with_bool"]) {
            sentence += "<strong>WITH</strong> ";
        } else {
            sentence += "<strong>WITHOUT</strong> ";
        }
        sentence += "<strong>" + params["field"] + "</strong> ";
        sentence += "between ";
        sentence += "<strong>" + params["value_bxy"][0] + "</strong> ";
        sentence += "and "
        sentence += "<strong>" + params["value_bxy"][1] + "</strong>.";
        
        return sentence;
    }

    var result = createObject(IntegerCountConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function IntegerAverageConstraint(shouldBool, field,
    fieldMin, fieldMax) {
    // Each team <should/shouldn't> have average <field> between
    // <min and <max>

    var isError = false;
    var errMsg = "";

    if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    } else if (!isNumber(fieldMin)) {
        errMsg = "Invalid minimum average";
        isError = true;
    } else if (!isNumber(fieldMax)) {
        errMsg = "Invalid maximum average";
        isError = true;
    } else if (fieldMin > fieldMax) {
        errMsg = "Min average cannot exceed max average"
        isError = true;
    }

    var params = {
        "constr_type" : "IntegerAverageConstraint",
        "name" : field + " constraint",
        "should_bool" : (shouldBool == 'true'),
        "field" : field,
        "priority" : 1,
        "average_bxy" : [fieldMin, fieldMax]
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        console.log(typeof params["should_bool"]);
        if (params["should_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "average "
        sentence += "<strong>" + params["field"] + "</strong>";
        sentence += " between <strong>" + params["average_bxy"][0] + "</strong>";
        sentence += " and <strong>" + params["average_bxy"][1] + "</strong>";

        return sentence;
    }

    var result = createObject(IntegerAverageConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function IntegerSimilarityConstraint(similarBool, field) {
    // Each team <should/shouldn't> have similar <field>

    var isError = false;
    var errMsg = "";

    if (!isBoolean(similarBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "IntegerSimilarityConstraint",
        "name" : field + " constraint",
        "similar_bool" : (similarBool == 'true'),
        "field" : field,
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["similar_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "similar "
        sentence += "<strong>" + params["field"] + "</strong>";

        return sentence;
    }

    var result = createObject(IntegerSimilarityConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function IntegerGlobalAverageConstraint(field) {
    // All teams should have similar <field>

    var isError = false;
    var errMsg = "";

    var params = {
        "constr_type" : "IntegerGlobalAverageConstraint",
        "name" : field + " constraint",
        "field" : field,
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>All groups have similar " + params["field"];
        return sentence;
    }

    var result = createObject(IntegerSimilarityConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function OptionCountConstraint(shouldBool, countMin, countMax,
    withBool, field, option) {
    // Each team <should/shouldn't> have between <min>
    // and <max> members <with/without> <field> equal to <option>

    var isError = false;
    var errMsg = "";

    if (!isNumber(countMin) || !isNumber(countMax) || countMin > countMax) {
        errMsg = "Invalid count constraint bounds";
        isError = true;
    } else if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "OptionCountConstraint",
        "name" : field + " constraint",
        "should_bool" : (shouldBool == 'true'),
        "count_bxy" : [countMin, countMax],
        "with_bool" : (withBool == 'true'),
        "field" : field,
        "selection" : option,
        "candidates": [],
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function setCandidates(candidates) {
        if (isArray(candidates)) {
            params.candidates = [...candidates];
        }
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["should_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "between ";
        sentence += "<strong>" + params["count_bxy"][0] + "</strong> ";
        sentence += "and "
        sentence += "<strong>" + params["count_bxy"][1] + "</strong> ";
        sentence += "members ";
        if (params["with_bool"]) {
            sentence += "<strong>WITH</strong> ";
        } else {
            sentence += "<strong>WITHOUT</strong> ";
        }
        sentence += "<strong>" + params["field"] + "</strong> ";
        sentence += "equal to ";
        sentence += "<strong>" + params["selection"] + "</strong> ";
        
        return sentence;
    }

    var result = createObject(OptionCountConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.setCandidates = setCandidates;
    result.getField = getField;

    return result;
}

function OptionSimilarityConstraint(shouldBool, field) {
    // Each team <should/shouldn't> have similar <field>

    var isError = false;
    var errMsg = "";

    if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "OptionSimilarityConstraint",
        "name" : field + " constraint",
        "similar_bool" : (shouldBool == 'true'),
        "field" : field,
        "candidates" : [],
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function setCandidates(candidates) {
        if (isArray(candidates)) {
            params.candidates = [...candidates];
        }
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["similar_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "similar "
        sentence += "<strong>" + params["field"] + "</strong>";

        return sentence;
    }

    var result = createObject(OptionSimilarityConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.setCandidates = setCandidates;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function SubsetCountConstraint(shouldBool, countMin, countMax,
    withBool, field, option) {
    // Each team <should/shouldn't> have between <min>
    // and <max> members <with/without> <field> containing <option>

    var isError = false;
    var errMsg = "";

    if (!isNumber(countMin) || !isNumber(countMax) || countMin > countMax) {
        errMsg = "Invalid count constraint bounds";
        isError = true;
    } else if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "SubsetCountConstraint",
        "name" : field + " constraint",
        "should_bool" : (shouldBool == 'true'),
        "count_bxy" : [countMin, countMax],
        "with_bool" : (withBool == 'true'),
        "field" : field,
        "selection" : option,
        "candidates": [],
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function setCandidates(candidates) {
        if (isArray(candidates)) {
            params.candidates = [...candidates];
        }
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["should_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "between ";
        sentence += "<strong>" + params["count_bxy"][0] + "</strong> ";
        sentence += "and "
        sentence += "<strong>" + params["count_bxy"][1] + "</strong> ";
        sentence += "members ";
        if (params["with_bool"]) {
            sentence += "<strong>WITH</strong> ";
        } else {
            sentence += "<strong>WITHOUT</strong> ";
        }
        sentence += "<strong>" + params["field"] + "</strong> ";
        sentence += "containing ";
        sentence += "<strong>" + params["selection"] + "</strong> ";
        
        return sentence;
    }

    var result = createObject(SubsetCountConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.setCandidates = setCandidates;
    result.getField = getField;

    return result;
}

function SubsetSimilarityConstraint(shouldBool, field) {
    // Each team <should/shouldn't> have similar <field>

    var isError = false;
    var errMsg = "";

    if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "SubsetSimilarityConstraint",
        "name" : field + " constraint",
        "similar_bool" : (shouldBool == 'true'),
        "field" : field,
        "candidates" : [],
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function setCandidates(candidates) {
        if (isArray(candidates)) {
            params.candidates = [...candidates];
        }
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["similar_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "similar "
        sentence += "<strong>" + params["field"] + "</strong>";

        return sentence;
    }

    var result = createObject(SubsetSimilarityConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.setCandidates = setCandidates;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.getField = getField;

    return result;
}

function BooleanCountConstraint(shouldBool, countMin, countMax,
    withBool, field) {
    // Each team <should/shouldn't> have between <min>
    // and <max> members <with/without> <field>

    var isError = false;
    var errMsg = "";

    if (!isNumber(countMin) || !isNumber(countMax) || countMin > countMax) {
        errMsg = "Invalid count constraint bounds";
        isError = true;
    } else if (!isBoolean(shouldBool)) {
        errMsg = "Invalid constraint boolean";
        isError = true;
    }

    var params = {
        "constr_type" : "BooleanCountConstraint",
        "name" : field + " constraint",
        "should_bool" : (shouldBool == 'true'),
        "count_bxy" : [countMin, countMax],
        "with_bool" : (withBool == 'true'),
        "field" : field,
        "priority" : 1
    }

    function getField() {
        return params["field"];
    }

    function hasError() {
        return isError;
    }

    function getErrorMsg() {
        return errMsg;
    }

    function setCandidates(candidates) {
        if (isArray(candidates)) {
            params.candidates = [...candidates];
        }
    }

    function toJSON() {
        return JSON.parse(JSON.stringify(params));
    }

    function toHTML() {
        var sentence = "<p>Each group ";
        if (params["should_bool"]) {
            sentence += "<strong>SHOULD HAVE</strong> ";
        } else {
            sentence += "<strong>SHOULDN'T HAVE</strong> ";
        }
        sentence += "between ";
        sentence += "<strong>" + params["count_bxy"][0] + "</strong> ";
        sentence += "and "
        sentence += "<strong>" + params["count_bxy"][1] + "</strong> ";
        sentence += "members ";
        if (params["with_bool"]) {
            sentence += "<strong>WITH</strong> ";
        } else {
            sentence += "<strong>WITHOUT</strong> ";
        }
        sentence += "<strong>" + params["field"] + "</strong> ";
        
        return sentence;
    }

    var result = createObject(OptionCountConstraint.prototype)
    result.toJSON = toJSON;
    result.toHTML = toHTML;
    result.hasError = hasError;
    result.getErrorMsg = getErrorMsg;
    result.setCandidates = setCandidates;
    result.getField = getField;

    return result;
}